While All/Any

TheAbsolutionism opened the pull request:
### Description
This PR aims to add AND conditions + OR conditions to SecWhile

---
**Target Minecraft Versions:** any <!-- 'any' means all supported versions -->
**Requirements:** none <!-- Required plugins, server software... -->
**Related Issues:** #5883 <!-- Links to related issues -->

Asleeepp commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -53,55 +66,143 @@
+	private enum WhileState{
+		NORMAL("[:do] while <.+>"),
+		ALL("[:do] while all"),
+		ANY("[:do] while (any|at least one of)"),
```
i don't feel `at least one of` makes much sense here
just `at least one` would make more sense, or some other alternative

TheAbsolutionism replied:
Thats how it is on SecConditional, I can just make it optional

APickledWalrus reviewed (changes requested):
Some early thoughts. I like the idea 🙂

Moderocky commented:
While this looks like a good start, I'm going to mark this as draft currently because there are some methodological things to think about (e.g. we have if any/all and while any/all, are there ways we can re-use parts of the multiline condition code? Are there other sections/conditions that might benefit from having something like an any/all? etc.) and it might be worth exploring this a bit more to see if there's anything fancy we can do :)

APickledWalrus commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -117,7 +222,16 @@ public TriggerItem getActualNext() {
 	public String toString(@Nullable Event event, boolean debug) {
-		return (doWhile ? "do " : "") + "while " + condition.toString(event, debug);
+		SyntaxStringBuilder builder = new SyntaxStringBuilder(event, debug);
+		if (doWhile && state != WhileState.DO)
```
```suggestion
		if (doWhile)
```
Technically not necessary since `state == WhileDate.DO` implies `!doWhile`

TheAbsolutionism replied:
Not necessarily, just as I transfer over the conditions from the  multilined all/any, I also transfer over the doWhile

APickledWalrus replied:
Ahh that's fair. Disregard this.

APickledWalrus commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -18,85 +18,190 @@
 
-	@SuppressWarnings("NotNullFieldNotInitialized")
-	private Condition condition;
+	private enum WhileState {NORMAL, ANY, DO}
```
```suggestion
	private enum WhileState {
		NORMAL, ANY, DO
	}
```

APickledWalrus commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -18,85 +18,190 @@
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
+		state = WHILE_STATE_PATTERNS.getInfo(matchedPattern);
 		doWhile = parseResult.hasTag("do");
+		multiline = parseResult.regexes.isEmpty() && state != WhileState.DO;
```
```suggestion
		multiline = matchedPattern == 1 || matchedPattern == 2;
```
I feel like this is a bit easier to understand

APickledWalrus commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -126,4 +240,31 @@ public void exit(Event event) {
 		super.exit(event);
 	}
 
+	private Conditional<Event> getConditional() {
+		return conditional;
+	}
+
+	private boolean isDoWhile() {
+		return doWhile;
+	}
```
```suggestion
	private Conditional<Event> getConditional() {
		return conditional;
	}

	private boolean isDoWhile() {
		return doWhile;
	}
```
Are these needed? I think you should be able to reference the fields directly.

APickledWalrus commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -126,4 +240,31 @@ public void exit(Event event) {
+		return doWhile;
+	}
+
+	private @Nullable Node getNextNode(Node precedingNode, ParserInstance parser) {
```
This appears to be able to be a static method

TheAbsolutionism replied:
I was actually working on somewhat adding this to `TriggerItems` 
or another class if you suggest

APickledWalrus replied:
You copied it from SecConditional right? Maybe it could go on the Node class? I'm not 100% sure.

APickledWalrus commented on src/main/java/ch/njol/skript/sections/SecWhile.java:
```diff
@@ -53,55 +64,136 @@
+						continue;
+					childKey = ScriptLoader.replaceOptions(childKey);
+					parser.setNode(childNode);
+					Condition condition1 = Condition.parse(childKey, "Can't understand the condition: '" + childKey + "'");
```
Unresolved? If it doesn't work, just let me know 🙂

APickledWalrus reviewed (changes requested):
Looks pretty good overall.

One (cursed) change we might want to make is with do while.
Rather than:
```
do while:
  a
  b
do:
  # code
```
I'd like to hear what others think.
We might have...
```
do:
  # code
while:
  a
  b
```

TheAbsolutionism commented:
@APickledWalrus Just wanted to touch base a little. 
For your suggestion, personally I'm not a fan of having the conditions being defined after the section to execute. Mainly because it doesn't fit. Implementing it wouldn't be hard. But say I have like 20 lines within the code to run, then the conditions afterward makes it seem awkward. 
However, I do understand the ambiguity of having the `[:do]` and the actual `do` section. Maybe we can just come up with an alternative syntax, preferably for the newly added `do` section to avoid breaking changes, or if you think having this minor breaking change is "ok" then that works too. Some alternatives I was thinking of are: run, execute, trigger. But obviously these are already used elsewhere, leaving words like: perform, process, start, operate, etc. Which some may be fine, others not so much.

Also, regarding @Moderocky 's comment. I would like to say I've fully met the conditions. As only some parts are reused between both SecWhile and SecConditional, and the rest are unique to their behavior. Whether or not the methods I added are good or not, well is up to the teams discretion. The methods I added should cover a general basis if any elements were capable of implementing an `any/all`, presuming it doesn't fall out of scope of these methods, which then would be unique to an element and/or if shared, then a new method could be added.

Anyways, that about sums it up. Let me know your thoughts.

sovdeeth commented:
I think 
```
do:
 ...
while:
 a
 b
``` is much more readable and intuitive than the current order, since it's read in order of execution
